Minseo

superozing opened the pull request.

superozing committed:
init commit

superozing committed:
setting commit

superozing committed:
#1 shortlink test

superozing committed:
init pseudo code

superozing committed:
#1 데이터 입력 구현
push_back, push_front, 생성자, 소멸자

jeontigger reviewed (commented):
good 좋지만 begin, end, erase도 구현 할거지?

superozing committed:
슈도코드 수정

superozing committed:
iterator 선언, begin, end, erase일부

superozing committed:
슈도코드 수정

jeontigger commented on Minseo/TList.h:
```diff
@@ -99,6 +140,36 @@ void TList<T>::push_front(const T& data)
 }
 
+template<typename T>
+typename TList<T>::iterator TList<T>::erase(iterator& iter)
+{
+	TNode<T>* NodeBuf = iter.m_pTarget->m_pNext;
+	// ���� ó��) ���� List ��ü�� �������ΰ�?
+	assert(iter.m_pOwner == this && iter.m_pTarget != nullptr);
+	if (iter.m_pTarget == m_pHead)
+	{
+		iter.m_pTarget->m_pNext->m_pPrev = nullptr;
+		m_pHead = m_pHead->m_pNext;
+	}
+	else
+	{
+		iter.m_pTarget->m_pNext->m_pPrev = iter.m_pTarget->m_pPrev;
+	}
+	if (iter.m_pTarget == m_pTail)
+	{
+		iter.m_pTarget->m_pPrev->m_pNext = nullptr;
+		m_pTail = m_pTail->m_pPrev;
+	}
+	else
+	{
+		iter.m_pTarget->m_pPrev->m_pNext = iter.m_pTarget->m_pNext;
+	}
+	delete iter.m_pTarget;
```
이럴땐 직관적인 이해를 돕기 위해 iter.m_pTarget을 하나의 Node<T>*로 스택 변수를 선언해서 사용하는게 좋아보임!

superozing replied:
@jeontigger iter.m_pTarget을 하나의 Node*로 스택 변수로 선언해서 사용하는게 Node*로 m_pTarget를 받는 변수를 만들어서 그걸로 사용하라는 뜻이야? 가독성을 위해서?

jeontigger replied:
가독성을 위해서 뿐만 아니라, 개발 속도도 향상 될 것으로 판단돼. iter.m_pTarget을 여러번 반복해서 쓰는 것 보다 Node* target = iter.m_pTarget; 으로 해서 target->prev->next; 이런식으로 사용하는게 더 좋을 것 같아

jeontigger replied:
erase 커밋에 잘 반영된듯!

superozing committed:
erase

superozing committed:
iterator함수와 연산자 오버로딩
++ -- 의 전위, 후위
==, !=
*를 사용해 iterator를 통한 값 참조

superozing committed:
#1 template linked list
생성자, 연산자 오버로딩 (이동)
swap 함수 - swap시 move를 사용해서 이동대입연산자 호출
insert 함수

superozing committed:
UTF-8 변환

superozing committed:
슈도 코드 수정

superozing committed:
#1 pop_front, pop_back, front, back

superozing committed:
#1 crtdbg.h 메모리 누수 검사

superozing committed:
UTF-8 변환

superozing committed:
파일 위치 수정, BST 파일 생성

superozing committed:
#8 슈도 코드 생성

일부 계획 새웠고 구현하고 계획 세우는 것을 반복 할 생각

jeontigger commented on Minseo/TList.h:
```diff
@@ -67,6 +68,68 @@ class TList
+			assert(!(m_pOwner == nullptr || m_pTarget == nullptr));
+			m_pTarget = m_pTarget->m_pNext;
+			return Buffer;
+		}
```
++()가 있으니까 재사용 하면 좋을듯!

jeontigger commented on Minseo/TList.h:
```diff
@@ -203,6 +209,19 @@ void TList<T>::push_front(const T& data)
+	Target->m_pNext = NewNode;
+	++m_Count;
+}
+
```
헤더랑 테일에 지정이 없는 것으로 보임!

superozing replied:
@jeontigger 실제 List쪽 구현이 어떤지 보지 않아서 모르지만 나는 iterator가 가리킨 다음 노드로 넣을 생각이야. 만약 타겟의 다음 노드가 테일이면 테일을 새로 만든 노드로 지정 해주어야 하는 것은 알겠는데 헤더랑 어떤 연관이 있는거야??

jeontigger replied:
현재 헤더의 앞쪽이거나 현재 테일의 뒷쪽에 삽입하는 경우의 처리가 안되어있다고 생각했고, 지금 봐도 그게 맞는 것 같아

jeontigger commented on Minseo/TList.h:
```diff
@@ -236,6 +255,38 @@ typename TList<T>::iterator TList<T>::erase(iterator& iter)
+	other.m_pHead = nullptr;
+	other.m_pTail = nullptr;
+	return *this;
+}
```
메모리 누수가 예상됨

jeontigger reviewed (commented):
전체적으로 읽기 편안함!

superozing committed:
증감연산자 후위 수정, 메모리 누수 방지, insert에서의 Tail 연결

- 기존 구현된 전위 버전 연산자를 이용해서 후위 버전이 깔끔하게 보임
- insert함수에서 Tail 노드가 이상해질 뻔한 상황을 수정
- 이동 대입 연산자에서 예외 처리

superozing committed:
#1 피드백 반영
증감연산자 후위 수정, 메모리 누수 방지, insert에서의 Tail 연결

superozing committed:
#8 insert 함수 구현

키 값이 같은 경우에는 해쉬맵처럼 덮어씌우도록...

superozing committed:
#8 UTF-8 변환

superozing committed:
#8 clear 구현

superozing committed:
#8 ++, --, erase, Has와 is함수

superozing committed:
#8 *, ->, ==, !=

superozing committed:
#8 오류 수정

실행하고 나서 보이는 오류 들을 고치고 make_pair 비슷한 함수도 하나 만들었어요.

superozing committed:
#8 생성자 오버로딩, begin, end

미처 구현하지 못 한 것들

superozing committed:
#1 이동 연산자 수정

이동 연산자 호출 시 메모리 누수가 일어나지 않도록 처음에 리스트 전체를 delete 하도록 수정했고, 이걸 그냥 함수로 빼서 소멸자에 넣었음
